src/serac/physics/solid_mechanics_contact.hpp

```diff
@@ -0,0 +1,353 @@
+   *
+   * @param interaction_id Unique identifier for the ContactInteraction
+   * @param bdry_attr_surf1 MFEM boundary attributes for the first surface
+   * @param bdry_attr_surf2 MFEM boundary attributes for the second surface
+   * @param contact_opts Defines contact method, enforcement, type, and penalty
+   */
+  void addContactInteraction(int interaction_id, const std::set<int>& bdry_attr_surf1,
+                             const std::set<int>& bdry_attr_surf2, ContactOptions contact_opts)
+  {
+    SLIC_ERROR_ROOT_IF(!is_quasistatic_, "Contact can only be applied to quasistatic problems.");
+    SLIC_ERROR_ROOT_IF(order_ > 1, "Contact can only be applied to linear (order = 1) meshes.");
+    contact_.addContactInteraction(interaction_id, bdry_attr_surf1, bdry_attr_surf2, contact_opts);
+  }
+
+  /**
+   * @brief Complete the initialization and allocation of the data structures.
+   *
+   * @note This must be called before AdvanceTimestep().
+   */
+  void completeSetup() override
+  {
+    int    cycle = 0;
+    double time  = 0.0;
+    double dt    = 0.0;
+    contact_.update(cycle, time, dt);
+
+    SolidMechanicsBase::completeSetup();
+  }
+
+  /// @brief Solve the Quasi-static Newton system
+  void quasiStaticSolve(double dt) override
+  {
+    time_ += dt;
+
+    // the ~85 lines of code below are essentially equivalent to the 1-liner
+    // u += dot(inv(J), dot(J_elim[:, dofs], (U(t + dt) - u)[dofs]));
+    // or, with Lagrange multiplier contact enforcement
+    // [u; p] += dot(inv(block_J), dot(block_J_elim[:, dofs], (U(t + dt) - u)[dofs]))
+    // where block_J = | J  B^T |
+    //                 | B   0  |
+
+    // Update the linearized Jacobian matrix
+    // In general, the solution vector is a stacked (block) vector:
+    //  | displacement     |
+    //  | contact pressure |
+    // Contact pressure is only active when solving a contact problem with Lagrange multipliers.
+    // The gradient is not a function of the Lagrange multipliers, so they do not need to be copied to the solution.
+    // However, the solution vector must be sized to the Operator, which includes the Lagrange multipliers.
+    mfem::Vector augmented_solution(displacement_.Size() + contact_.numPressureDofs());
+    augmented_solution = 0.0;
+    augmented_solution.SetVector(displacement_, 0);
+    residual_with_bcs_->GetGradient(augmented_solution);
+
+    du_ = 0.0;
+    for (auto& bc : bcs_.essentials()) {
+      bc.setDofs(du_, time_);
+    }
+
+    auto& constrained_dofs = bcs_.allEssentialTrueDofs();
+    for (int i = 0; i < constrained_dofs.Size(); i++) {
+      int j = constrained_dofs[i];
+      du_[j] -= displacement_(j);
+    }
+
+    dr_ = 0.0;
+    mfem::EliminateBC(*J_, *J_e_, constrained_dofs, du_, dr_);
+
+    // Update the initial guess for changes in the parameters if this is not the first solve
+    for (std::size_t parameter_index = 0; parameter_index < parameters_.size(); ++parameter_index) {
```

Comment:
Yes, it _should_ be the same.  The problem is changing the size of the linear solver's operator to do the warm start on just the displacement dofs then changing it back to include the Lagrange multiplier's blocks was giving me an error with SuperLU.  To get around this, I do the warm start with contact including the Lagrange multipliers.

I don't know enough about SuperLU to know if changing the operator size triggering an error is the appropriate behavior and it doesn't look like we can copy the linear solver easily.  Any thoughts on how this might be handled better @jamiebramwell?

Reply:
We talked about this offline, but I've noticed odd memory issues with SuperLU. I recommend trying Strumpack to see if that resolves the issue.

Reply:
Fixed now that we're using Strumpack!